File: android/truepad-spt/src/main/kotlin/dev/systemslibrarian/truepad/spt/ComparisonWords.kt
package dev.systemslibrarian.truepad.spt

/* ============================================================================
 * The comparison wordlist — the BIP-39 English list, vendored.
 * ----------------------------------------------------------------------------
 * THESE ARE NOT MNEMONICS. TruePad uses this list as a human-readable encoding
 * of fixed 11-bit indices and nothing else. It does not create BIP-39 wallet
 * mnemonics, apply BIP-39 checksum rules, derive seeds, or accept a phrase
 * typed by a user. Nothing here is key material; the words exist so two people
 * can say twelve or eight things aloud and notice if they differ.
 *
 * INDEX POSITION IS THE PROTOCOL MAPPING. The confirmation ceremony compares
 * the SAME words on both devices, so the Kotlin list must equal the Browser
 * Edition's COMPARISON_WORDS line for line. It is loaded from the vendored
 * `comparison-words.txt` resource (byte-identical to src/browser/ui/wordlist/
 * english.txt), and ComparisonWordsTest pins its SHA-256 and shape so a build
 * that ever diverges fails rather than silently disagreeing about a fingerprint.
 *
 * Provenance, integrity and licence: see COMPARISON-WORDS-PROVENANCE.md beside
 * the resource (MIT; upstream bitcoin/bips bip-0039/english.txt).
 * ========================================================================= */

/** The 2048-entry comparison wordlist, index-addressed. Loaded once, verified
 *  in shape at load so a corrupt or wrong-length resource fails closed. */
object ComparisonWords {

    /** Resource path, relative to this class's package on the classpath. */
    private const val RESOURCE = "/comparison-words.txt"

    /** Exactly 2048 words: an 11-bit index space (the fingerprints emit 11-bit
     *  indices), index 0 = "abandon" … index 2047 = "zoo". */
    const val COUNT: Int = 2048

    val words: List<String> by lazy { load() }

    private fun load(): List<String> {
        val stream = ComparisonWords::class.java.getResourceAsStream(RESOURCE)
            ?: error("comparison wordlist resource $RESOURCE is missing from the build")
        val lines = stream.bufferedReader(Charsets.UTF_8).use { it.readText() }
            .split('\n')
            // The vendored file ends with a trailing newline, so the split yields
            // one empty tail element; drop only a single trailing empty entry.
            .let { if (it.isNotEmpty() && it.last().isEmpty()) it.dropLast(1) else it }
        require(lines.size == COUNT) { "comparison wordlist must hold $COUNT words, found ${lines.size}" }
        return lines
    }

    /** The word for one 11-bit index. Refuses an out-of-range index rather than
     *  wrapping or clamping — a bad index is a protocol error, not a display one. */
    fun wordAt(index: Int): String {
        require(index in 0 until COUNT) { "comparison index $index is outside 0..${COUNT - 1}" }
        return words[index]
    }

    /** Render a run of indices (the twelve request words, the eight confirmation
     *  words) as their words, in order — order is part of the ceremony. */
    fun wordsFor(indices: IntArray): List<String> = indices.map { wordAt(it) }
}

File: ios/TruePadKit/Sources/TruePadSPT/ComparisonWords.swift
import Foundation

/* ============================================================================
 * The comparison wordlist — the BIP-39 English list, vendored.
 *
 * THESE ARE NOT MNEMONICS. TruePad uses this list as a human-readable encoding of
 * fixed 11-bit indices and nothing else. It does not create BIP-39 wallet
 * mnemonics, apply BIP-39 checksum rules, derive seeds, or accept a phrase typed
 * by a user. NOTHING HERE IS KEY MATERIAL. The words exist so that two people can
 * say twelve or eight things aloud and notice if they differ.
 *
 * INDEX POSITION IS THE PROTOCOL MAPPING. The confirmation ceremony compares the
 * SAME words on both devices, so this list must equal the Browser and Android
 * Editions' line for line. The resource is byte-identical to
 * src/browser/ui/wordlist/english.txt and android's comparison-words.txt, and
 * ComparisonWordsTests pins its SHA-256 — the same value Android pins — so a
 * build that ever diverges FAILS rather than silently disagreeing about a
 * fingerprint two people are reading to each other.
 *
 * Provenance, integrity and licence: see COMPARISON-WORDS-PROVENANCE.md beside
 * the resource (MIT; upstream bitcoin/bips bip-0039/english.txt).
 * ========================================================================= */

public enum ComparisonWords {
    /// Exactly 2048 words: an 11-bit index space, because the fingerprints emit
    /// 11-bit indices. Index 0 is "abandon"; index 2047 is "zoo".
    public static let count = 2048

    public enum Failure: Error, Equatable {
        case resourceMissing
        case wrongLength(Int)
        case malformedEntry(Int)
    }

    /// The list, loaded once and verified in SHAPE at load, so a corrupt or
    /// wrong-length resource fails closed rather than producing short phrases.
    public static let words: [String] = {
        guard let loaded = try? load() else {
            // A build without the resource cannot do the ceremony at all, and
            // must not fall back to anything. An empty list makes every lookup
            // refuse rather than return a plausible wrong word.
            return []
        }
        return loaded
    }()

    static func load() throws -> [String] {
        guard let url = Bundle.module.url(forResource: "comparison-words", withExtension: "txt"),
              let text = try? String(contentsOf: url, encoding: .utf8) else {
            throw Failure.resourceMissing
        }
        let lines = text.split(separator: "\n", omittingEmptySubsequences: false)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        guard lines.count == count else { throw Failure.wrongLength(lines.count) }
        for (i, word) in lines.enumerated() {
            // ASCII lowercase letters only. Anything else means the resource was
            // re-encoded or edited, and two devices could then read different
            // things aloud while believing they agreed.
            guard !word.isEmpty, word.allSatisfy({ $0.isASCII && $0.isLowercase && $0.isLetter }) else {
                throw Failure.malformedEntry(i)
            }
        }
        return lines
    }

    /// The words for a list of 11-bit indices.
    ///
    /// Returns nil if ANY index is out of range or the list failed to load —
    /// never a partial phrase. A ceremony with a missing word is a ceremony where
    /// one side reads eleven and the other twelve, and notices too late.
    public static func render(_ indices: [Int]) -> [String]? {
        let list = words
        guard list.count == count else { return nil }
        var out: [String] = []
        out.reserveCapacity(indices.count)
        for index in indices {
            guard index >= 0, index < count else { return nil }
            out.append(list[index])
        }
        return out
    }
}
